### PR TITLE
KEYCLOAK-6262 Incorporate new visual design from PatternFly

### PR DESCRIPTION
**Pull request by junezhang**
Currently, UXD team finished the new login design in PatternFly. http://www.patternfly.org/pattern-library/application-framework/login-page/

### This PR is for
-  incorporating the new visual design from PatternFly to KeyCloak.

#### Basic Login Page (no social provider)
screen shot 2018-02-07 at 6 30 08 pm

Responsive Design (on mobile)
screen shot 2018-02-07 at 6 32 37 pm

#### Login with social provider
One column (when the amount is less or equal 4)
screen shot 2018-02-07 at 6 35 54 pm

Two columns (when the amount is more than 4)
screen shot 2018-02-07 at 6 34 59 pm

Responsive Design (on mobile)
screen shot 2018-02-07 at 6 36 05 pm

A gif to check what the new login looks like

### CODE REVIEW

File: themes/src/main/resources/theme/base/login/template.ftl [hunk 1]
```diff
@@ -32,62 +32,55 @@
...
-                                <#list locale.supported as l>
-                                    <li class="kc-dropdown-item"><a href="${l.url}">${l.label}</a></li>
-                                </#list>
-                            </ul>
-                        </div>
+  <div class="${properties.kcLoginClass!}">
+    <header class="${properties.kcHeaderClass!}">
+      <div id="kc-logo"><a href="${properties.kcLogoLink!'#'}"><div id="kc-logo-wrapper"></div></a></div>
+    </header>
+    <div class="${properties.kcFormCardClass!} <#if realm.password && social.providers??>${properties.kcFormCardAccountClass!}</#if>">
```
Comment (stianst) [inline themes/src/main/resources/theme/base/login/template.ftl]: This is causing a problem on some pages as "realm.password" and "social.providers" are not available on all pages. I don't quite understand what this is doing and why it adds this class like this. Can you explain?

**Review by stianst — DISMISSED** [review]
Found a few broken things:

* Login page shows "Username" when it should show "Username or email"
* Forgot password page shows "Username or email" above input instead of inside
* Title on all pages is "Login In" (password reset page, register page, update password page, etc.)
* Grant page is broken
* Configure OTP page is broken
* OTP page is broken
* IdP Link to existing account is broken
* Terms and conditions page is broken
* Error page is completely broken due to "if realm.password .." added to template.ftl

File: themes/src/main/resources/theme/base/login/template.ftl (hunk 1 above)
Comment (junezhang) [reply themes/src/main/resources/theme/base/login/template.ftl]: Sure, @stianst. The logical of this piece of code is the login box width is different, when no social provide, the width is 500px. and when adding the social provider the width of login box is 840px.
As here, I need a class to control the width.

Comment (stianst) [reply themes/src/main/resources/theme/base/login/template.ftl]: https://github.com/junezhang/keycloak/pull/1

**Comment by junezhang** [comment]
@stianst Updating as following:

1, Update the username placeholder to 
`<#if !realm.loginWithEmailAllowed>${msg("username")}<#elseif !realm.registrationEmailAsUsername>${msg("usernameOrEmail")}<#else>${msg("email")}</#if>`
2, Update the Forgot password page shows "Username or email" above input instead of inside.
same with the above method. 
3, Update the Grant Page
screen shot 2018-02-12 at 8 28 51 pm
4, Update the Term Page
screen shot 2018-02-12 at 8 28 45 pm

As in my local server, when I open "Update password" "Config OTP", it leads to an error page.
The server error log is here
https://gist.github.com/junezhang/205a03c286e1464ab1ee3624679a1ef5
https://gist.github.com/anonymous/918e1f5e1eb4f8aaf6a4b16c71a80047

Please help to check it.
Thanks.

**Comment by stianst** [comment]
One more thing we need to consider. We have the Realm HTML name which allows users to show a name or a image for the realm. This is now completely gone and the realm name doesn't exist anywhere. That means a user doesn't know what realm they are authenticating.

**Comment by junezhang** [comment]
@stianst Have fixed the Realm name and image as the attachment.

Realm with image


Realm with text

**Comment by junezhang** [comment]
@stianst @ssilvert 
I've fixed the "Config OTP", "Update Password", "Update profile".

**Comment by stianst** [comment]
Fixed some minor margin/padding issue in otp config screens:
https://github.com/keycloak/keycloak/pull/4983/commits/e36dfcc1406e0d382ddf2912700bfcdbe3a7cd5d

**Comment by stianst** [comment]
There's two outstanding issues:

* Title is "Log In" on all pages. This doesn't make any sense. We should probably change this to have the realm name on the top always (above the white box) and replace "Log In" with a page specific title.

* Weird title "MY REALM ACCOUNT." on consent screen. This is broken in master as well. The fix for the above would help here. Keep "MY REALM" at the top and replace "Log In" with something like "Grant access to App Name".

**Review by stianst — CHANGES_REQUESTED** [review]
There's two outstanding issues:

* Title is "Log In" on all pages. This doesn't make any sense. We should probably change this to have the realm name on the top always (above the white box) and replace "Log In" with a page specific title.

* Weird title "MY REALM ACCOUNT." on consent screen. This is broken in master as well. The fix for the above would help here. Keep "MY REALM" at the top and replace "Log In" with something like "Grant access to App Name".

**Comment by stianst** [comment]
We need to review RH-SSO theme and maybe also update RCUE styles

**Review by stianst — CHANGES_REQUESTED** [review]
We need to review RH-SSO theme and maybe also update RCUE styles

**Comment by stianst** [comment]
image

RH-SSO theme is completely broken, so that needs looking into.

**Comment by junezhang** [comment]
@stianst The RH-SSO theme look like below:
screen shot 2018-03-14 at 6 12 25 pm
How do you think if we add the Red Hat logo on the top-right?

**Comment by stianst** [comment]
Why did you add themes/src/main/resources-product/theme/rh-sso/login/resources/css/login.css? I deliberately removed it and added login-rhsso.css instead so it can inherit styles from Keycloak theme instead of duplicating it. We've had quite a few issues here in the past where the two have been inconsitent.

**Comment by junezhang** [comment]
@stianst Removed the login.css from resources-product. please have a review.

**Comment by seb54000** [comment]
I think the minus sign in the login.css file for display higher than 768px is a caps
https://github.com/keycloak/keycloak/blob/190771ddf18d89a3242737d533a01b48bec82556/themes/src/main/resources-product/theme/rh-sso/login/resources/css/login.css#L272 

This lead to a behaviour of getting locale listbox (for internationalization) out of the screen

**Comment by junezhang** [comment]
@seb54000 Thanks for your feedback.

In our latest login design, we fixed this issue. 
https://github.com/junezhang/keycloak/blob/c8e6a9947e28a3a2385dea9bf41b11f201658867/themes/src/main/resources/theme/keycloak/login/resources/css/login.css

The screenshot is as below


mobile screenshot
screen shot 2018-03-15 at 5 51 27 pm

Please help to check. thanks.

**Comment by stianst** [comment]
A lot of tests are failing. This is primarily due to the changes in the page title. I'm currently working on updating the tests and will send a commit when I have tests passing.

**Comment by seb54000** [comment]
@junezhang Great ! :raised_hands:

**Comment by stianst** [comment]
@junezhang squashed the commits into a single commit and rebased from master so it should now be ready to merge once @vmuzikar has reviewed

**Review by stianst — DISMISSED** [review]

**Comment by vmuzikar** [comment]
I've tried to thoroughly test the new login screen in all supported browsers (both mobile and desktop) and found a few issues (their priority/severity is probably a subject to discuss).

**Overflowing text for some social provider buttons**
This is happening when the browser window is a bit narrow but not enough for switching to the mobile version. This is mainly a problem for tablets in portrait mode.
Also IdPs with long names don't have a proper padding.
screen shot 2018-03-16 at 11 23 10

**Overflowing realm name**
Long realm names are overflowing and covering localization dropdown on smartphones.
screen shot 2018-03-16 at 11 27 05

**Placeholders don't work while focused in IE**
When you select a text input field which has a placeholder (the labeling text), the placeholder disappears even though the field is blank (i.e. no text was entered yet).
This could be an UX problem mainly in combination with autofocus which we use and could the user confuse a bit. Take the following 2 scenarios:
- User opens the login page and sees two fields from which one is just blank (with no hint if they should enter username, email, ...).
- User clicks "Forgot Password?" on the login page and sees just one blank input field with no clue what should they enter.
capture

**Console UI Tests are broken**
Some of them are failing in IE and FF, all of them in Chrome.
I can fix them but it won't be sooner than Monday (and possibly even Tuesday).

@stianst WDYT? How severe do you consider the issues? Should we merge it now or should we at least fix the tests? I'd prefer the latter.

**Comment by junezhang** [comment]
@vmuzikar Thanks for checking.

Fixed the overflowing text for some social provider buttons
screen shot 2018-03-19 at 4 04 34 pm

Fixed overflowing realm name
screen shot 2018-03-19 at 4 05 25 pm

Currently, in my local server, I don't have IE, so can't test it.

**Comment by vmuzikar** [comment]
@junezhang Thanks for the fixes!

I believe I've discovered another one.
On Android phones it seems like the IdP buttons which have no icon (e.g. GitLab) are very "grayed out" and even a different font (larger size?).
screen shot 2018-03-19 at 12 34 30
However, it might be an issue specific to the Android Emulator I use. We need to try to reproduce it with a real Android phone.

**Comment by vmuzikar** [comment]
@junezhang I've fixed the Console UI Tests and created a PR for it: https://github.com/junezhang/keycloak/pull/2

**Comment by vmuzikar** [comment]
Also, I can confirm that the issue described in previous comment (https://github.com/keycloak/keycloak/pull/4983#issuecomment-374186101) is present with real Android phones as well.

**Comment by junezhang** [comment]
@vmuzikar fixed the different style in different Idp buttons. Below is the screenshots of the emulated device - Nexus. please help to check on your Android phone, if it is right?
screen shot 2018-03-20 at 12 25 15 pm

**Comment by vmuzikar** [comment]
@junezhang Thanks for the fixes.

I can confirm that the IdP buttons now look good on Android phones.

However, I'm not sure about adding the labels back. Imho it doesn't look quite "elegant" having both placeholders and labels. @stianst WDYT?

screen shot 2018-03-20 at 10 16 40

I see 3 options here.
1. Use labels instead placeholders in the IE solely
1. Use some different approach for placeholders which would achieve their full functionality in the IE. That would, however, require Javascript and that makes it not a usable solution for us, I suppose.
1. Use only the placeholders and just ignore a bit worse UX in the IE.

Also, we need to fix the issue with the StackOverflow button on tablets - https://github.com/keycloak/keycloak/pull/4983#issuecomment-374181605.

**Comment by vmuzikar** [comment]
@stianst Should we include some Red Hat logo in the RH-SSO theme?

**Comment by stianst** [comment]
@vmuzikar there's a separate JIRA for updating the logo

**Comment by stianst** [comment]
I would remove the placeholders completely. We don't want to introduce JS in the login frames and as such there's no way to properly support IE. It also makes it consistent with other pages.

@junezhang I believe you may have just forgotten to remove the placeholders after re-introducing the labels?

**Comment by vmuzikar** [comment]
@stianst +1
Even though the placeholders looked very cool, clean and modern (I really liked them), labels are the only solution how to provide seamless and functional design across all the browsers.

**Comment by stianst** [comment]
We can probably revisit labels later

**Review by stianst — DISMISSED** [review]

**Comment by stianst** [comment]
Removed the placeholders. Also fixed the Keycloak background which wasn't showing properly on wide-screens.

**Review by vmuzikar — APPROVED** [review]

**Review by stianst — APPROVED** [review]

**Pull request merged**